resolve/metadata: Stop encoding macros as reexports

petrochenkov opened the pull request:
Supersedes https://github.com/rust-lang/rust/pull/88335.
r? @cjgillot

petrochenkov committed:
resolve: Fix incorrect results of `opt_def_kind` query for some built-in macros

Previously it always returned `MacroKind::Bang` while some of those macros are actually attributes and derives

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-12`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/4493479054?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/6e41171c6db505b47894e4b063632a27a1af62ab/checks/4493479054/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
    Checking core v0.0.0 (/checkout/library/core)
   Compiling compiler_builtins v0.1.55
    Checking rustc-std-workspace-core v1.99.0 (/checkout/library/rustc-std-workspace-core)
 Documenting alloc v0.0.0 (/checkout/library/alloc)
error: `core::cmp::Ord` is both a trait and a macro
    |
    |
254 | /// [`Ord`]: core::cmp::Ord
    |              ^^^^^^^^^^^^^^ ambiguous link
    |
    = note: `-D rustdoc::broken-intra-doc-links` implied by `-D warnings`
help: to link to the trait, prefix with `trait@`
    |
254 | /// [`Ord`]: trait@core::cmp::Ord
    |              ++++++
help: to link to the macro, add an exclamation mark
    |
254 | /// [`Ord`]: core::cmp::Ord!


error: `core::cmp::Ord` is both a trait and a macro
   |
   |
30 | /// [`Ord`]: core::cmp::Ord
   |              ^^^^^^^^^^^^^^ ambiguous link
   |
help: to link to the trait, prefix with `trait@`
   |
30 | /// [`Ord`]: trait@core::cmp::Ord
   |              ++++++
help: to link to the macro, add an exclamation mark
   |
30 | /// [`Ord`]: core::cmp::Ord!


error: `core::cmp::Ord` is both a trait and a macro
     |
     |
2802 | /// Implements ordering of vectors, [lexicographically](core::cmp::Ord#lexicographical-comparison).
     |                                                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ ambiguous link
     |
help: to link to the trait, prefix with `trait@`
     |
2802 | /// Implements ordering of vectors, [lexicographically](trait@core::cmp::Ord#lexicographical-comparison).
     |                                                         ++++++
help: to link to the macro, add an exclamation mark
     |
2802 | /// Implements ordering of vectors, [lexicographically](core::cmp::Ord#lexicographical-comparison!).


error: `core::cmp::Ord` is both a trait and a macro
     |
     |
2790 | /// Implements comparison of vectors, [lexicographically](core::cmp::Ord#lexicographical-comparison).
     |                                                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ ambiguous link
     |
help: to link to the trait, prefix with `trait@`
     |
2790 | /// Implements comparison of vectors, [lexicographically](trait@core::cmp::Ord#lexicographical-comparison).
     |                                                           ++++++
help: to link to the macro, add an exclamation mark
     |
2790 | /// Implements comparison of vectors, [lexicographically](core::cmp::Ord#lexicographical-comparison!).

error: could not document `alloc`

Caused by:
Caused by:
  process didn't exit successfully: `/checkout/obj/build/bootstrap/debug/rustdoc --edition=2018 --crate-type lib --crate-name alloc library/alloc/src/lib.rs --target x86_64-unknown-linux-gnu -o /checkout/obj/build/x86_64-unknown-linux-gnu/stage2-std/x86_64-unknown-linux-gnu/doc --cfg 'feature="compiler-builtins-c"' --error-format=json --json=diagnostic-rendered-ansi --markdown-css rust.css --markdown-no-toc -Z unstable-options --resource-suffix 1.59.0 --index-page /checkout/src/doc/index.md -C metadata=2b278315e6dda61e -L dependency=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-std/x86_64-unknown-linux-gnu/release/deps -L dependency=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-std/release/deps --extern compiler_builtins=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-std/x86_64-unknown-linux-gnu/release/deps/libcompiler_builtins-2c9c844da8a1e155.rmeta --extern core=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-std/x86_64-unknown-linux-gnu/release/deps/libcore-24cf655a97be028a.rmeta -Zsymbol-mangling-version=legacy -Dwarnings '-Wrustdoc::invalid_codeblock_attributes' --crate-version '1.59.0-nightly
  (d684a5624
  2021-12-11)' '-Zcrate-attr=doc(html_root_url="https://doc.rust-lang.org/nightly/")'` (exit status: 1)


command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "rustdoc" "--target" "x86_64-unknown-linux-gnu" "-Zbinary-dep-depinfo" "-j" "16" "--release" "--locked" "--color" "always" "--features" "panic-unwind backtrace compiler-builtins-c" "--manifest-path" "/checkout/library/test/Cargo.toml" "-p" "alloc" "-Zskip-rustdoc-fingerprint" "--" "--markdown-css" "rust.css" "--markdown-no-toc" "-Z" "unstable-options" "--resource-suffix" "1.59.0" "--index-page" "/checkout/src/doc/index.md"


Build completed unsuccessfully in 0:25:44
```

</details>

bors commented:
:umbrella: The latest upstream changes (presumably #91769) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-12`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/4498283321?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/1b373717de694028e3a9409dec3766360dcd9bf2/checks/4498283321/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Set({"src/doc/edition-guide"}) not skipped for "bootstrap::doc::EditionGuide" -- not in ["src/tools/tidy"]
Rustbook (x86_64-unknown-linux-gnu) - edition-guide
Building stage0 tool linkchecker (x86_64-unknown-linux-gnu)
    Finished release [optimized] target(s) in 0.13s
std/convert/trait.AsRef.html:12: broken link - `std/convert/crate::cmp::Eq`
std/convert/trait.AsRef.html:12: broken link - `std/convert/crate::cmp::Ord`
std/cmp/index.html:6: broken intra-doc link - [<code>Eq</code>] and [<code>PartialEq</code>]
std/cmp/index.html:9: broken intra-doc link - [<code>Ord</code>] and [<code>PartialOrd</code>]
std/cmp/index.html:12: broken intra-doc link - [<code>Ord</code>]
std/cmp/index.html:13: broken intra-doc link - [<code>PartialOrd</code>]
std/cmp/index.html:15: broken intra-doc link - [<code>Ord</code>]
std/cmp/trait.Ord.html:9: broken intra-doc link - [<code>PartialOrd</code>]
std/cmp/trait.Ord.html:52: broken intra-doc link - [<code>PartialOrd</code>] and [<code>Eq</code>] (which requires [<code>PartialEq</code>]
std/cmp/trait.Eq.html:11: broken intra-doc link - [<code>PartialEq</code>]
std/cmp/trait.PartialOrd.html:26: broken intra-doc link - [<code>Ord</code>]
std/cmp/trait.PartialOrd.html:53: broken intra-doc link - [<code>PartialEq</code>]
std/cmp/trait.PartialOrd.html:54: broken intra-doc link - [<code>Ord</code>]
std/cmp/trait.PartialEq.html:18: broken intra-doc link - [<code>PartialOrd</code>] or [<code>Ord</code>]
std/iter/trait.Iterator.html:424: broken link - `std/iter/crate::clone::Clone`
std/iter/trait.Iterator.html:450: broken link - `std/iter/crate::clone::Clone`
alloc/sync/struct.Arc.html:324: broken link - `core/intrinsics/%7B%7Bmisc%7D%7D/fn.transmute.html`
alloc/rc/struct.Rc.html:220: broken link - `core/intrinsics/%7B%7Bmisc%7D%7D/fn.transmute.html`
number of HTML files scanned: 31799
number of HTML redirects found: 9842
number of links checked: 2153322
number of links ignored due to external: 90574
```

</details>

petrochenkov commented:
Blocked on https://github.com/rust-lang/rust/pull/92034 and https://github.com/rust-lang/rust/pull/92086.

bors commented:
:umbrella: The latest upstream changes (presumably #92153) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-12`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/4760694764?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/377606298dd17bf90de547d845855a04dfa812f3/checks/4760694764/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Set({"src/doc/edition-guide"}) not skipped for "bootstrap::doc::EditionGuide" -- not in ["src/tools/tidy"]
Rustbook (x86_64-unknown-linux-gnu) - edition-guide
Building stage0 tool linkchecker (x86_64-unknown-linux-gnu)
    Finished release [optimized] target(s) in 0.15s
std/iter/trait.Iterator.html:426: broken link - `std/iter/crate::clone::Clone`
std/iter/trait.Iterator.html:452: broken link - `std/iter/crate::clone::Clone`
std/convert/trait.AsRef.html:14: broken link - `std/convert/crate::cmp::Eq`
std/convert/trait.AsRef.html:14: broken link - `std/convert/crate::cmp::Ord`
std/cmp/trait.PartialEq.html:20: broken intra-doc link - [<code>PartialOrd</code>] or [<code>Ord</code>]
std/cmp/trait.PartialOrd.html:28: broken intra-doc link - [<code>Ord</code>]
std/cmp/trait.PartialOrd.html:55: broken intra-doc link - [<code>PartialEq</code>]
std/cmp/trait.PartialOrd.html:56: broken intra-doc link - [<code>Ord</code>]
std/cmp/trait.Eq.html:13: broken intra-doc link - [<code>PartialEq</code>]
std/cmp/trait.Ord.html:11: broken intra-doc link - [<code>PartialOrd</code>]
std/cmp/trait.Ord.html:54: broken intra-doc link - [<code>PartialOrd</code>] and [<code>Eq</code>] (which requires [<code>PartialEq</code>]
std/cmp/index.html:8: broken intra-doc link - [<code>Eq</code>] and [<code>PartialEq</code>]
std/cmp/index.html:11: broken intra-doc link - [<code>Ord</code>] and [<code>PartialOrd</code>]
std/cmp/index.html:14: broken intra-doc link - [<code>Ord</code>]
std/cmp/index.html:15: broken intra-doc link - [<code>PartialOrd</code>]
std/cmp/index.html:17: broken intra-doc link - [<code>Ord</code>]
number of HTML files scanned: 31973
number of HTML redirects found: 9914
number of links checked: 2158826
number of links ignored due to external: 89386
```

</details>

bors commented:
:umbrella: The latest upstream changes (presumably #92970) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

bors commented:
:umbrella: The latest upstream changes (presumably #93245) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

petrochenkov committed:
metadata: Tweak the way in which declarative macros are encoded

To make the `macro_rules` flag more readily available without decoding everything else

petrochenkov committed:
resolve/metadata: Stop encoding macros as reexports

petrochenkov commented:
This should be ready now.
@bors ready

cjgillot reviewed (commented):
I understand that this PR resolves the ambiguity between normal exports and `macro_rules` re-export by adding a `macro_rules` flag to metadata to distinguish between the two. In that case, I'm not sure to understand what is the benefit compared to the former setup.
Did I miss something?

petrochenkov commented on compiler/rustc_middle/src/hir/map/mod.rs:
```diff
@@ -229,7 +229,14 @@ impl<'hir> Map<'hir> {
+                        .macro_kind_overrides
+                        .get(&local_def_id)
+                        .copied()
+                        .unwrap_or(MacroKind::Bang),
```
Yes, I think that would be better.
I assumed that with something like https://github.com/rust-lang/rust/pull/90451 it doesn't matter because the tables used by lowering will still need to live in `ResolverOutputs`, but if there are multiple resolver output structures then there's still a potential to put this table to some more short living structure.

petrochenkov commented on compiler/rustc_resolve/src/build_reduced_graph.rs:
```diff
@@ -1221,6 +1221,9 @@ impl<'a, 'b> BuildReducedGraphVisitor<'a, 'b> {
         let (ext, ident, span, macro_rules) = match &item.kind {
             ItemKind::MacroDef(def) => {
                 let ext = Lrc::new(self.r.compile_macro(item, self.r.session.edition()));
+                if let macro_kind @ (MacroKind::Attr | MacroKind::Derive) = ext.macro_kind() {
```
Maybe, I'll try to put it there but there shouldn't be much difference.

petrochenkov commented:
>I understand that this PR resolves the ambiguity between normal exports and `macro_rules` re-export by adding a `macro_rules` flag to metadata to distinguish between the two. In that case, I'm not sure to understand what is the benefit compared to the former setup.

Previously macros were encoded as *both* items and reexports, but the item versions were ignored during decoding to avoid duplication.
Now they are only encoded as items are decoded as items, which is more right because macros are not reexports and it was strange to encode them as such, and also this allows to just encode less data.

petrochenkov commented:
For performance it would be optimal to use two structures:
```rust
// This struct is used during decoding and unifies both proper items and reexports
struct ModChild {
    pub ident: Ident,
    pub res: Res<!>,
    pub vis: ty::Visibility,
    pub span: Span,
    pub macro_rules: bool,
}

// This structure is used for encoding the reexport table
pub struct Reexport {
    pub ident: Ident,
    pub res: Res<!>,
    pub vis: ty::Visibility,
    pub span: Span,
}
```
, but I just didn't want to introduce more entities.
I can try it in a separate PR, but the perf benefits from not encoding this extra `bool` will probably be lost in noise, so it's likely not worth it.

petrochenkov commented:
Updated.

rust-log-analyzer commented:
The job **`x86_64-gnu-tools`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/5069631011?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/63b50ec86241dbf4073bdd6fb23a54e398518941/checks/5069631011/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
  IMAGE: x86_64-gnu-tools
##[endgroup]
From https://github.com/rust-lang/rust
 * branch              master     -> FETCH_HEAD
Searching for toolstate changes between 4e8fb743ccbec27344b2dd42de7057f41d4ebfdd and ee3cfa1518001d9e25caf128e366ab51dd3881ca
Clippy or rustfmt subtrees were updated
##[group]Run src/ci/scripts/verify-channel.sh
src/ci/scripts/verify-channel.sh
shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}
env:
---
To only update this specific test, also pass `--test-args fs.rs`

error: 2 errors occurred comparing output.
status: exit status: 1
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools-bin/miri" "tests/run-pass/fs.rs" "-L" "/tmp/compiletestSia442" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "-C" "prefer-dynamic" "-o" "/tmp/compiletestSia442/fs.stage-id" "-A" "unused" "--edition" "2018" "-Astable-features" "--sysroot" "/home/user/.cache/miri/HOST" "-Zmiri-disable-isolation" "-L" "/tmp/compiletestSia442/fs.stage-id.aux"
------------------------------------------

------------------------------------------
stderr:
---



The actual stderr differed from the expected stderr.
Actual stderr saved to /tmp/compiletestSia442/portable-simd.stage-id.stderr
To only update this specific test, also pass `--test-args portable-simd.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
status: exit status: 1
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools-bin/miri" "tests/run-pass/portable-simd.rs" "-L" "/tmp/compiletestSia442" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "-C" "prefer-dynamic" "-o" "/tmp/compiletestSia442/portable-simd.stage-id" "-A" "unused" "--edition" "2018" "-Astable-features" "--sysroot" "/home/user/.cache/miri/HOST" "-L" "/tmp/compiletestSia442/portable-simd.stage-id.aux"
------------------------------------------

------------------------------------------
stderr:
---

 error: `macro_use` attributes are no longer needed in the Rust 2018 edition
   --> $DIR/macro_use_imports.rs:18:5
    |
 LL |     #[macro_use]
-   |     ^^^^^^^^^^^^ help: remove the attribute and import the macro directly, try: `use mac::{pub_macro, inner_mod_macro, function_macro, ty_macro, pub_in_private_macro};`
+   |     ^^^^^^^^^^^^ help: remove the attribute and import the macro directly, try: `use mac::{pub_macro, function_macro, ty_macro, inner_mod_macro, pub_in_private_macro};`
    |
    = note: `-D clippy::macro-use-imports` implied by `-D warnings`
 error: `macro_use` attributes are no longer needed in the Rust 2018 edition
   --> $DIR/macro_use_imports.rs:20:5
    |
 LL |     #[macro_use]
 LL |     #[macro_use]
    |     ^^^^^^^^^^^^ help: remove the attribute and import the macro directly, try: `use mini_mac::ClippyMiniMacroTest;`
 error: `macro_use` attributes are no longer needed in the Rust 2018 edition
   --> $DIR/macro_use_imports.rs:22:5
    |
 LL |     #[macro_use]
 LL |     #[macro_use]
    |     ^^^^^^^^^^^^ help: remove the attribute and import the macro directly, try: `use mac::{inner::foofoo, inner::try_err};`
 error: `macro_use` attributes are no longer needed in the Rust 2018 edition
   --> $DIR/macro_use_imports.rs:24:5
    |
 LL |     #[macro_use]
 LL |     #[macro_use]
    |     ^^^^^^^^^^^^ help: remove the attribute and import the macro directly, try: `use mac::inner::nested::string_add;`
 error: aborting due to 4 previous errors
 
 


The actual stderr differed from the expected stderr.
Actual stderr saved to /checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/test/macro_use_imports.stage-id.stderr
diff of fixed:

 // aux-build:macro_rules.rs
 // aux-build:macro_use_helper.rs
 // aux-build:proc_macro_derive.rs
 // run-rustfix
 // ignore-32bit
 
 #![allow(unused_imports, unreachable_code, unused_variables, dead_code, unused_attributes)]
 #![allow(clippy::single_component_path_imports)]
 #![warn(clippy::macro_use_imports)]
 #[macro_use]
 extern crate macro_use_helper as mac;
 
 #[macro_use]
 #[macro_use]
 extern crate proc_macro_derive as mini_mac;
 
 mod a {
-    use mac::{pub_macro, inner_mod_macro, function_macro, ty_macro, pub_in_private_macro};
+    use mac::{pub_macro, function_macro, ty_macro, inner_mod_macro, pub_in_private_macro};
     use mac;
     use mini_mac::ClippyMiniMacroTest;
     use mini_mac;
     use mac::{inner::foofoo, inner::try_err};
     use mac::inner;
     use mac::inner::nested::string_add;
     use mac::inner::nested;
 
     #[derive(ClippyMiniMacroTest)]
     struct Test;
     fn test() {
         pub_macro!();
         pub_macro!();
         inner_mod_macro!();
         pub_in_private_macro!(_var);
         function_macro!();
         let v: ty_macro!() = Vec::default();
         inner::try_err!();
         inner::try_err!();
         inner::foofoo!();
         nested::string_add!();
 }
 
 
 // issue #7015, ICE due to calling `module_children` with local `DefId`
 use a as b;
 
 fn main() {}
 
 

The actual fixed differed from the expected fixed.
Actual fixed saved to /checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/test/macro_use_imports.stage-id.fixed
To only update this specific test, also pass `--test-args macro_use_imports.rs`

error: 2 errors occurred comparing output.
status: exit status: 1
status: exit status: 1
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/clippy-driver" "tests/ui/macro_use_imports.rs" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/test" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "-C" "prefer-dynamic" "-o" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/test/macro_use_imports.stage-id" "-A" "unused" "--emit=metadata" "-Dwarnings" "-Zui-testing" "-L" "dependency=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps" "-L" "dependency=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/release/deps" "--extern" "itertools=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libitertools-041fb6ac880e1ce0.rlib" "--extern" "syn=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libsyn-c2a1ca34edf818c9.rlib" "--extern" "derive_new=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/release/deps/libderive_new-12319133577eb155.so" "--extern" "serde=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libserde-a8c1b2a71f554c3c.rlib" "--extern" "if_chain=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libif_chain-a436811527635382.rlib" "--extern" "parking_lot=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libparking_lot-e6439823b6d16f2c.rlib" "--extern" "regex=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libregex-8ccd5459decf8e02.rlib" "--extern" "quote=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libquote-a2cb7849bbc8a2a2.rlib" "--extern" "futures=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libfutures-b682a5a8a9c64b20.rlib" "--extern" "tokio=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libtokio-6688bfb3a0699fc9.rlib" "--extern" "serde_derive=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/release/deps/libserde_derive-f91723cecf6d8a5f.so" "--extern" "clippy_utils=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libclippy_utils-84536a848ae0c873.rlib" "--edition=2021" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/test/macro_use_imports.stage-id.aux"
------------------------------------------

------------------------------------------
stderr:
stderr:
------------------------------------------
{"message":"`macro_use` attributes are no longer needed in the Rust 2018 edition","code":{"code":"clippy::macro_use_imports","explanation":null},"level":"error","spans":[{"file_name":"tests/ui/macro_use_imports.rs","byte_start":427,"byte_end":439,"line_start":18,"line_end":18,"column_start":5,"column_end":17,"is_primary":true,"text":[{"text":"    #[macro_use]","highlight_start":5,"highlight_end":17}],"label":null,"suggested_replacement":null,"suggestion_applicability":null,"expansion":null}],"children":[{"message":"`-D clippy::macro-use-imports` implied by `-D warnings`","code":null,"level":"note","spans":[],"children":[],"rendered":null},{"message":"remove the attribute and import the macro directly, try","code":null,"level":"help","spans":[{"file_name":"tests/ui/macro_use_imports.rs","byte_start":427,"byte_end":439,"line_start":18,"line_end":18,"column_start":5,"column_end":17,"is_primary":true,"text":[{"text":"    #[macro_use]","highlight_start":5,"highlight_end":17}],"label":null,"suggested_replacement":"use mac::{pub_macro, function_macro, ty_macro, inner_mod_macro, pub_in_private_macro};","suggestion_applicability":"MaybeIncorrect","expansion":null}],"children":[],"rendered":null}],"rendered":"error: `macro_use` attributes are no longer needed in the Rust 2018 edition\n  --> tests/ui/macro_use_imports.rs:18:5\n   |\nLL |     #[macro_use]\n   |     ^^^^^^^^^^^^ help: remove the attribute and import the macro directly, try: `use mac::{pub_macro, function_macro, ty_macro, inner_mod_macro, pub_in_private_macro};`\n   |\n   = note: `-D clippy::macro-use-imports` implied by `-D warnings`\n\n"}
{"message":"`macro_use` attributes are no longer needed in the Rust 2018 edition","code":{"code":"clippy::macro_use_imports","explanation":null},"level":"error","spans":[{"file_name":"tests/ui/macro_use_imports.rs","byte_start":457,"byte_end":469,"line_start":20,"line_end":20,"column_start":5,"column_end":17,"is_primary":true,"text":[{"text":"    #[macro_use]","highlight_start":5,"highlight_end":17}],"label":null,"suggested_replacement":null,"suggestion_applicability":null,"expansion":null}],"children":[{"message":"remove the attribute and import the macro directly, try","code":null,"level":"help","spans":[{"file_name":"tests/ui/macro_use_imports.rs","byte_start":457,"byte_end":469,"line_start":20,"line_end":20,"column_start":5,"column_end":17,"is_primary":true,"text":[{"text":"    #[macro_use]","highlight_start":5,"highlight_end":17}],"label":null,"suggested_replacement":"use mini_mac::ClippyMiniMacroTest;","suggestion_applicability":"MaybeIncorrect","expansion":null}],"children":[],"rendered":null}],"rendered":"error: `macro_use` attributes are no longer needed in the Rust 2018 edition\n  --> tests/ui/macro_use_imports.rs:20:5\n   |\nLL |     #[macro_use]\n   |     ^^^^^^^^^^^^ help: remove the attribute and import the macro directly, try: `use mini_mac::ClippyMiniMacroTest;`\n\n"}
{"message":"`macro_use` attributes are no longer needed in the Rust 2018 edition","code":{"code":"clippy::macro_use_imports","explanation":null},"level":"error","spans":[{"file_name":"tests/ui/macro_use_imports.rs","byte_start":492,"byte_end":504,"line_start":22,"line_end":22,"column_start":5,"column_end":17,"is_primary":true,"text":[{"text":"    #[macro_use]","highlight_start":5,"highlight_end":17}],"label":null,"suggested_replacement":null,"suggestion_applicability":null,"expansion":null}],"children":[{"message":"remove the attribute and import the macro directly, try","code":null,"level":"help","spans":[{"file_name":"tests/ui/macro_use_imports.rs","byte_start":492,"byte_end":504,"line_start":22,"line_end":22,"column_start":5,"column_end":17,"is_primary":true,"text":[{"text":"    #[macro_use]","highlight_start":5,"highlight_end":17}],"label":null,"suggested_replacement":"use mac::{inner::foofoo, inner::try_err};","suggestion_applicability":"MaybeIncorrect","expansion":null}],"children":[],"rendered":null}],"rendered":"error: `macro_use` attributes are no longer needed in the Rust 2018 edition\n  --> tests/ui/macro_use_imports.rs:22:5\n   |\nLL |     #[macro_use]\n   |     ^^^^^^^^^^^^ help: remove the attribute and import the macro directly, try: `use mac::{inner::foofoo, inner::try_err};`\n\n"}
{"message":"`macro_use` attributes are no longer needed in the Rust 2018 edition","code":{"code":"clippy::macro_use_imports","explanation":null},"level":"error","spans":[{"file_name":"tests/ui/macro_use_imports.rs","byte_start":529,"byte_end":541,"line_start":24,"line_end":24,"column_start":5,"column_end":17,"is_primary":true,"text":[{"text":"    #[macro_use]","highlight_start":5,"highlight_end":17}],"label":null,"suggested_replacement":null,"suggestion_applicability":null,"expansion":null}],"children":[{"message":"remove the attribute and import the macro directly, try","code":null,"level":"help","spans":[{"file_name":"tests/ui/macro_use_imports.rs","byte_start":529,"byte_end":541,"line_start":24,"line_end":24,"column_start":5,"column_end":17,"is_primary":true,"text":[{"text":"    #[macro_use]","highlight_start":5,"highlight_end":17}],"label":null,"suggested_replacement":"use mac::inner::nested::string_add;","suggestion_applicability":"MaybeIncorrect","expansion":null}],"children":[],"rendered":null}],"rendered":"error: `macro_use` attributes are no longer needed in the Rust 2018 edition\n  --> tests/ui/macro_use_imports.rs:24:5\n   |\nLL |     #[macro_use]\n   |     ^^^^^^^^^^^^ help: remove the attribute and import the macro directly, try: `use mac::inner::nested::string_add;`\n\n"}

------------------------------------------

thread 'compile_test' panicked at 'Some tests failed', /cargo/registry/src/github.com-1ecc6299db9ec823/compiletest_rs-0.7.1/src/lib.rs:105:22
```

</details>

petrochenkov committed:
Update clippy tests

bors commented:
:umbrella: The latest upstream changes (presumably #94129) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

cjgillot commented:
Sorry for the delay.
r=me once rebased.

petrochenkov commented:
@bors r=cjgillot

bors commented:
:pushpin: Commit b91ec30159594ed73cabc4c6cb5450996f290577 has been approved by `cjgillot`

<!-- @bors r=cjgillot b91ec30159594ed73cabc4c6cb5450996f290577 -->
<!-- homu: {"type":"Approved","sha":"b91ec30159594ed73cabc4c6cb5450996f290577","approver":"cjgillot"} -->
